// react-style-singleton@2.2.3/dist/es2015/singleton.js
var r = Object.defineProperty;
var n = (t, e) => r(t, "name", { value: e, configurable: !0 });
import { getNonce as l } from "../../../../../get-nonce@1.0.1/node_modules/get-nonce/dist/es2015/index.js";
function a() {
  if (!document)
    return null;
  var t = document.createElement("style");
  t.type = "text/css";
  var e = l();
  return e && t.setAttribute("nonce", e), t;
}
n(a, "makeStyleTag");
function i(t, e) {
  t.styleSheet ? t.styleSheet.cssText = e : t.appendChild(document.createTextNode(e));
}
n(i, "injectStyles");
function c(t) {
  var e = document.head || document.getElementsByTagName("head")[0];
  e.appendChild(t);
}
n(c, "insertStyleTag");
var s = /* @__PURE__ */ n(function() {
  var t = 0, e = null;
  return {
    add: /* @__PURE__ */ n(function(o) {
      t == 0 && (e = a()) && (i(e, o), c(e)), t++;
    }, "add"),
    remove: /* @__PURE__ */ n(function() {
      t--, !t && e && (e.parentNode && e.parentNode.removeChild(e), e = null);
    }, "remove")
  };
}, "stylesheetSingleton");
export {
  s as stylesheetSingleton
};
